### PR TITLE
feat: add kdl

### PR DESCRIPTION
**Pull request by amaanq**
https://github.com/kdl-org/kdl/blob/main/SPEC.md

Grammar/Parser: https://github.com/amaanq/tree-sitter-kdl

~~Not sure why, but whoever requested it in #2282 seems to have deleted their comment, which is unfortunate.~~
cc: @MultisampledNight since they requested it first in #2282
(dunno why I couldn't find the comment before, oops)

Only thing missing is having the special slashdash comments actually be commented out. Still working on that as this was a pain to begin with.

I would also like someone to go over my `@number/@float` queries, float is embedded into number so I'm not sure if I separated them properly.

Thanks!
-Amaan

### CODE REVIEW

**Comment by amaanq** [comment]
Force pushed since I added readme

**Comment by amaanq** [comment]
Ok, I'm working out comments and might've run into a bug. Draft until fixed

**Comment by amaanq** [comment]
fixed the bug, now it's just comments that are the issue. here's several types of comments with the slash dash (/-) in kdl

image

the first one should have the whole block commented out, i.e. till the }. In fact, showing captures under cursor does show comment is there, but other highlights are overriding it, how can we either a) set conditionals on those to not highlight on comment, or b) override them with comment

the second one should have only /-key=1 commented, but "foo" got captured too since props or values after a /- can be commented, how can we get it to only pick 1 or the other? and same issue with =1 where other highlights take over

third one same as second

the last one should have everything from { } (inclusive) commented, captures does show its a comment, but again other hls take over

Thanks to anyone that can help me out!

**Comment by MultisampledNight** [comment]
<details>
<summary>(fwiw, here is an equivalent text version of that file)</summary>

```kdl
// entire node stuff commented out
/-this "stuff" a=1 {
  node
  in
  a
  node
}

// property is commented out
also /-a=1 "stuff" {
  node
  in
  a
  node
}

// name is commented out
but /-"stuff" a=1 {
  node
  in
  a
  node
}

// children are commented out
actually "stuff" a=1 /- {
  node
  in
  a
  node
}
```

<3

</summary>

**Comment by MultisampledNight** [comment]
Found that anonymous nodes might be interesting. The first case works slightly better with this query:

```
(node (node_slash_dash) (_)? @comment) @comment
```

(question mark only added for purity, has no effect here)

unknown

**Comment by MultisampledNight** [comment]
...wow, okay, nevermind. That gives really inconsistent results I don't quite grasp.

**Comment by MultisampledNight** [comment]
Seems like it's possible to configure priorities for queries, found that through https://github.com/nvim-treesitter/nvim-treesitter/issues/3058:

```
(node (node_slash_dash) (set! "priority" 105)) @comment
```

unknown

**Comment by theHamsta** [comment]
You seem to be using a scanner.c. Even though, you're not doing something complicated there. You could use https://github.com/vigoux/tree-sitter-fuzz-action in your repo

**Comment by amaanq** [comment]
> You seem to be using a scanner.c. Even though, you're not doing something complicated there. You could use https://github.com/vigoux/tree-sitter-fuzz-action in your repo

I had an issue detecting EOFs without just checking for 0 beforehand in the scanner. This helped alleviate a bug where tree-sitter would parse indefinitely at the end.

**Review by theHamsta — APPROVED** [review]

**Comment by amaanq** [comment]
> Seems like it's possible to configure priorities for queries, found that through https://github.com/nvim-treesitter/nvim-treesitter/issues/3058 (after looking a bit more around, seems like that's been added in https://github.com/neovim/neovim/pull/15114 and is documented in `treesitter.txt` under `treesitter-highlight-priority`):
> 
> ```
> (node (node_slash_dash) (#set! "priority" 105)) @comment
> ```
> 
> unknown
> 

have you had success with this? I'll play around with it soon so I can update the PR asap

**Comment by MultisampledNight** [comment]
Partially. It worked fine for the whole node and children slashdash comments, but not for the individual props/args (I guess `value` needs some further refinement since `prop` also contains one).

This is what I've got:

```diff
diff --git a/queries/kdl/highlights.scm b/queries/kdl/highlights.scm
index e83f4731..79dc53e8 100644
--- a/queries/kdl/highlights.scm
+++ b/queries/kdl/highlights.scm
@@ -50,7 +50,7 @@
   (multi_line_comment)
 ] @comment
 
-(node (node_slash_dash)) @comment
+(node (node_slash_dash) (#set! "priority" 105)) @comment
 (node (node_prop_or_arg_slash_dash) @comment (prop) @comment)
 (node (node_prop_or_arg_slash_dash) @comment (value) @comment)
-(node_children (node_children_slash_dash)) @comment
+(node_children (node_children_slash_dash) (#set! "priority" 105)) @comment

```

**Comment by amaanq** [comment]
Ok, I worked around KDL's IDL a bit and got it to work.. I guess rules are meant to be bent a bit

Here's sample output on my end now that I'll update v shortly
image

**Comment by amaanq** [comment]
unsure about folds since I've always been confused on whether or not say a fold marked @fold

foo {
  .....
}

should include or exclude the {} itself

Otherwise this is good to squash n' merge now @theHamsta

**Comment by MultisampledNight** [comment]
@amaanq I found `:help treesitter-highlight-priority`, which says (shortened):

> Treesitter uses `nvim_buf_set_extmark()` to set highlights with a default priority of 100. This enables plugins to set a highlighting priority lower or higher than tree-sitter.

As such, I believe higher means more important, lower means less important, and all other rules had a priority equal to the default of 100.t

**Comment by amaanq** [comment]
> @amaanq I found `:help treesitter-highlight-priority`, which says (shortened):
> 
> > Treesitter uses `nvim_buf_set_extmark()` to set highlights with a default priority of 100. This enables plugins to set a highlighting priority lower or higher than tree-sitter.
> 
> As such, I believe higher means more important, lower means less important, and all other rules had a priority equal to the default of 100.

Thanks! What do you think of the folds? I don't use folds personally so it'd be nice to hear from you

**Comment by MultisampledNight** [comment]
I don't use them either, personally. But I'll set them up and test.

**Comment by amaanq** [comment]
more of a TODO, but I'm thinking (maybe for the future I'm still relatively new to this) that we have uncommented stuff in a capture without comments, since anything commented should not have the normal highlight to begin with, then commented stuff is a pure comment capture, and then add injections to comments (I think that's their purpose)

**Comment by MultisampledNight** [comment]
It works okay, apart from a bit of weirdness I don't think you can do something about. Take this for example:

```kdl
but /-"stuff" a=1 {
  in
  a
  node {
    nestification {
      wow "a"
    }
  }
}
```

- `za` on `node` closes the children of `but`.
- `za` on `nestification`/`wow` closes the children of `nestification`, but leaves over a closing `}`.
- `za` on the closing `}` of `nestification` closes the children of `nestification`, with the `}` this time.
- Doesn't seem like I can close `node` only.

**Comment by amaanq** [comment]
Thanks to @ahlinc [here](https://github.com/amaanq/tree-sitter-kdl/pull/2), improvements have been made but as a result there's still a bug or two to be fixed when using kdl's examples instead of their tests. Namely this fails:

```kdl
// Nodes can be separated into multiple lines
title \
  "Some title"


// Files must be utf8 encoded!
smile "😁"

// Instead of anonymous nodes, nodes and properties can be wrapped
// in "" for arbitrary node names.
"!@#$@$%Q#$%~@!40" "1.2.3" "!!!!!"=true

// The following is a legal bare identifier:
foo123~!@#$%^&*.:'|?+ "weeee"

// And you can also use unicode!
ノード　お名前="☜(ﾟヮﾟ☜)"

// kdl specifically allows properties and values to be
// interspersed with each other, much like CLI commands.
foo bar=true "baz" quux=false 1 2 3
```

Draft until fixed

**Comment by amaanq** [comment]
Good to merge now, did a lot of thorough testing

File: queries/kdl/folds.scm [hunk 1]
```diff
@@ -0,0 +1,3 @@
+; Folds
+
+(node children: (node_children) @fold)
```
Comment (theHamsta) [inline queries/kdl/folds.scm:3]: Just `(node_children) @fold` would e equivalent?

**Review by theHamsta — APPROVED** [review]

File: queries/kdl/folds.scm (hunk 1 above)
Comment (amaanq) [reply queries/kdl/folds.scm]: yep you're right, done

**Comment by amaanq** [comment]
@theHamsta do you think this is good to merge now? Or just waiting for someone elses input?

**Comment by theHamsta** [comment]
Thank you!

**Comment by amaanq** [comment]
Thank you too! Glad to be a contributor :smile:

**Pull request merged**